lib/tree.rb

```diff
@@ -2,65 +2,150 @@ class TreeNode
+    #         current = current.left
+    #       else
+    #         current.left = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     else
+    #       if !current.right.nil?
+    #         current = current.right
+    #       else
+    #         current.right = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     end
+    #   end
+    # end
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    # base case when the current node is nil
+    return TreeNode.new(key, value) if current_node.nil?
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end
+
+  # Time Complexity:o(log n)
+  # Space Complexity:o(n)
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
+  end
+
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+    if key < current_node.key
+      return find_helper(current_node.left, key)
+    elsif key > current_node.key
+      return find_helper(current_node.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def inorder
-    raise NotImplementedError
+    inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+
+    return list
+  end
+
+  # Time Complexity: o(n)
+  # Space Complexity:o(n)
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
+  end
+
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def height
-    raise NotImplementedError
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value }
+    return list
   end
 
-  # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
-  def bfs
-    raise NotImplementedError
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
+  def height
```

Comment:
The space complexity is O(log n) if the tree is balanced and O(n) if it's not.
The method however works